test/funnels/Swapper.t.sol

```diff
@@ -0,0 +1,168 @@
+
+    address constant FACILITATOR = address(0x1337);
+    address constant KEEPER      = address(0xb0b);
+
+    uint8 constant SWAPPER_ROLE = uint8(1);
+
+    function setUp() public {
+        vm.createSelectFork(vm.envString("ETH_RPC_URL"));
+
+        buffer = new AllocatorBuffer();
+        roles = new AllocatorRoles();
+        swapper = new Swapper(address(roles), ilk, address(buffer));
+        uniV3Callee = new UniV3SwapperCallee(UNIV3_ROUTER);
+
+        roles.setIlkAdmin(ilk, address(this));
+        roles.setRoleAction(ilk, SWAPPER_ROLE, address(swapper), swapper.swap.selector, true);
+        roles.setUserRole(ilk, FACILITATOR, SWAPPER_ROLE, true);
+
+        swapper.setLimits(DAI, USDC, 3600 seconds, uint128(10_000 * WAD));
+        swapper.setLimits(USDC, DAI, 3600 seconds, uint128(10_000 * 10**6));
+
+        deal(DAI,  address(buffer), 1_000_000 * WAD,   true);
+        deal(USDC, address(buffer), 1_000_000 * 10**6, true);
+        buffer.approve(USDC, address(swapper), type(uint256).max);
+        buffer.approve(DAI,  address(swapper), type(uint256).max);
+    }
+
+    function testConstructor() public {
+        Swapper s = new Swapper(address(0xBEEF), "SubDAO 1", address(0xAAA));
+        assertEq(address(s.roles()),  address(0xBEEF));
+        assertEq(s.ilk(), "SubDAO 1");
+        assertEq(s.buffer(), address(0xAAA));
+        assertEq(s.wards(address(this)), 1);
+    }
+
+    function testAuth() public {
+        checkAuth(address(swapper), "Swapper");
+    }
+
+    function testModifiers() public {
+        bytes4[] memory authedMethods = new bytes4[](2);
+        authedMethods[0] = swapper.setLimits.selector;
+        authedMethods[1] = swapper.swap.selector;
+
+        vm.startPrank(address(0xBEEF));
+        checkModifier(address(swapper), "Swapper/not-authorized", authedMethods);
+        vm.stopPrank();
+    }
+
+    function testSetLimits() public {
+        vm.expectEmit(true, true, true, true);
+        emit SetLimits(address(1), address(2), 3, 4);
+        vm.prank(address(this)); swapper.setLimits(address(1), address(2), 3, 4);
+        (uint64 hop, uint64 zzz, uint128 cap) = swapper.limits(address(1), address(2));
+        assertEq(hop, 3);
+        assertEq(zzz, 0);
+        assertEq(cap, 4);
+    }
+
+    function testRoles() public {
+        vm.expectRevert("Swapper/not-authorized");
+        vm.prank(address(0xBEEF)); swapper.setLimits(address(0), address(0), 0, 0);
+        roles.setRoleAction(ilk, uint8(0xF1), address(swapper), bytes4(keccak256("setLimits(address,address,uint64,uint128)")), true);
+        roles.setUserRole(ilk, address(0xBEEF), uint8(0xF1), true);
+        vm.prank(address(0xBEEF)); swapper.setLimits(address(0), address(0), 0, 0);
+    }
+
+    function testSwap() public {
+        uint256 prevSrc = GemLike(USDC).balanceOf(address(buffer));
+        uint256 prevDst = GemLike(DAI).balanceOf(address(buffer));
+
+        vm.expectEmit(true, true, true, false);
```

Comment:
One way we could check the `out` value is actually executing it twice and reverting the first using vm.snapshot.

Reply:
Done, but currently not covering Depositor.withdraw as it logs `collected0` and `collected1`, which are not returned.
Need to decide if we want to return them.
I played a bit with it and got a bunch of stack too deep problems when returning them and testing the events in the same way.